stats/opentelemetry/grpc_trace_bin_propagator_test.go

```diff
@@ -0,0 +1,250 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package opentelemetry
+
+import (
+	"context"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	oteltrace "go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/metadata"
+	itracing "google.golang.org/grpc/stats/opentelemetry/internal/tracing"
+)
+
+var validSpanContext = oteltrace.SpanContext{}.WithTraceID(
+	oteltrace.TraceID{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}).WithSpanID(
+	oteltrace.SpanID{17, 18, 19, 20, 21, 22, 23, 24}).WithTraceFlags(
+	oteltrace.TraceFlags(1))
+
+// TestInject verifies that the GRPCTraceBinPropagator correctly injects
+// existing binary trace data or OpenTelemetry span context as `grpc-trace-bin`
+// header in the provided carrier's context metadata.
+//
+// It verifies that if a valid span context is injected, same span context can
+// can be retreived from the carrier's context metadata.
+//
+// If an invalid span context is injected, it verifies that `grpc-trace-bin`
+// header is not set in the carrier's context metadata.
+func (s) TestInject(t *testing.T) {
+	tests := []struct {
+		name     string
+		injectSC oteltrace.SpanContext
+		wantSC   oteltrace.SpanContext
+		md       metadata.MD
+		wantKeys []string
+	}{
+		{
+			name:     "valid OpenTelemetry span context",
+			injectSC: validSpanContext,
+			wantSC:   validSpanContext,
+			md:       metadata.MD{"incoming-key": []string{"incoming-value"}, "outgoing-key": []string{"outgoing-value"}},
+			wantKeys: []string{grpcTraceBinHeaderKey, "outgoing-key"},
+		},
+		{
+			name:     "invalid OpenTelemetry span context",
+			injectSC: oteltrace.SpanContext{},
+			wantSC:   oteltrace.SpanContext{},
+			md:       metadata.MD{"incoming-key": []string{"incoming-value"}, "outgoing-key": []string{"outgoing-value"}},
+			wantKeys: []string{"outgoing-key"},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			p := GRPCTraceBinPropagator{}
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			ctx = oteltrace.ContextWithSpanContext(ctx, test.injectSC)
+			ctx = metadata.NewIncomingContext(ctx, metadata.MD{"incoming-key": test.md["incoming-key"]})
+			ctx = metadata.NewOutgoingContext(ctx, metadata.MD{"outgoing-key": test.md["outgoing-key"]})
+
+			c := itracing.NewOutgoingCarrier(ctx)
+			p.Inject(ctx, c)
+
+			if gotKeys := c.Keys(); !cmp.Equal(test.wantKeys, gotKeys, cmpopts.SortSlices(func(a, b string) bool { return a < b })) {
+				t.Fatalf("c.Keys() = keys %v, want %v", gotKeys, test.wantKeys)
+			}
+			md, _ := metadata.FromOutgoingContext(c.Context())
+			gotH := md.Get(grpcTraceBinHeaderKey)
+			if !test.wantSC.IsValid() {
+				if len(gotH) > 0 {
+					t.Fatalf("got non-empty value from Carrier's context metadata grpc-trace-bin header, want empty")
```

Comment:
It's better to log what was observed, not just "non-empty".

Reply:
Added gotH to error message